### DISCUSSION TITLE
export mesh to Cesium (glTF/glb)

### DISCUSSION THREAD

**Discussion by leomiquelutti** [pyvista/pyvista#5627] [Q&A] [answered]
## Description

Hi everyone,

I need to export the meshes generated in `pyvista` to add them in [Resium](https://resium.darwineducation.com/) (the React [Cesium](https://cesium.com/index.html) component). AFAIK, 3D models formats accepted are `glTF` or `glb` (glTF binary form)

So far, I am working with `pyvista.StructuredGrid` mesh type (I am not sure if the class I am using makes any difference).

Would someone know how to do it, please?

**Comment by akaszynski** [comment]
Hello,

This is a feature we can add to `pyvista` provided you're using `vtk>=9.0.0`.  We can leverage the new feature from `vtk` using:
```python
import vtk
exporter= vtk.vtkGLTFExporter()
```

Would you be willing on trying this out on your end and then prototyping the code using `vtk` first?  We can then add that to our `fileio`.

**Comment by banesullivan** [comment, marked as answer]
We may add later at one point. I've been working with Cesium lately and use this code to export meshes to the `glb` format:

You'll need to install [`pygltflib`](https://pypi.org/project/pygltflib/)

```py
import pyvista as pv
import os, shutil
import vtk
from pygltflib.utils import gltf2glb


def export(mesh, basename):
    """Export a single mesh to glb format.
    
    Check results at https://gltf-viewer.donmccurdy.com/
    """
    p = pv.Plotter(notebook=False, off_screen=True) # might not always work
    p.add_mesh(mesh.extract_geometry(), color='white', lighting=False)
    p.show(auto_close=False)

    if os.path.exists('export/'):
        shutil.rmtree("export/")
    os.makedirs("export/")
    exp = vtk.vtkGLTFExporter()
    exp.SetInput(p.ren_win)
    exp.SetActiveRenderer(p.renderer)
    exp.SetFileName("export/{}.gltf".format(basename))
    exp.Write()

    p.close()


    gltf2glb("export/{}.gltf".format(basename))
    shutil.move("export/{}.glb".format(basename), "{}.glb".format(basename))
    shutil.rmtree("export/")
    return True
```

Example usage:

```py
from pyvista import examples
export(examples.load_structured(), 'wave')
```

then check the result here: https://gltf-viewer.donmccurdy.com/

**Comment by leomiquelutti** [comment]
Hi @banesullivan, thanks for this very useful code snippet!

It worked! At least where I was supposed to check the results (gltf-viewer). With `cmap` as argument instead of color! 🥳 

I will try it on Cesium asap. Once I do it, I'll post the outcome here.

**Comment by leomiquelutti** [comment]
Hi @banesullivan, 

the model was imported in Cesium partially successfull.  With both Cesium and [donmccurdy](https://gltf-viewer.donmccurdy.com/) viewer, the model was shown in an upward position.
![image](https://user-images.githubusercontent.com/9113903/89213654-cc275b80-d59b-11ea-8f34-70aab2860ab7.png)

We were able to solve that in Cesium, but I don't know if that is expected behavior. The mesh in `pyvista` lied in the horizontal position (as a map).

And in Cesium we solved that importing it as a `3DTileset` instead of a `3DModel`, which I personally believe was not the correct way to do so. We couldn't import it as a `3DModel`. However, we will sort that by ourselves with the Cesium support. 

If you know the reason why the model showed in the upright position and/or that has something to do with `pyvista`, let me know, please!

And thanks again! Saved me many coding hours. 😄

**Comment by banesullivan** [comment]
I've dealt with the same issue and I always just apply a transform in Cesium. It could be an issue in `vtk.vtkGLTFExporter` or it could be an issue in how Cesium imports those data. I have no idea. I just apply a transform to get what I expect 🤷‍♂️  and haven't had a serious project or effort that's made it worth exploring what's really going on there

**Comment by leomiquelutti** [comment]
Hi @banesullivan 

do you know why does [my model](https://drive.google.com/file/d/15Au2Nsq0yAlGf-JejNF_JzowtpNcSBbM/view?usp=sharing) get _invisible_ when upside down? This is preventing me from correct positioning the model on the globe. 

I am creating the 3D model this way:
```
grid = xr.open_dataset(gridName) # netCDF file
x = np.tile(grid.easting.values, (len(grid.northing.values),1))
y = np.tile(grid.northing.values, (len(grid.easting.values),1))
pvgrid = pv.StructuredGrid(x, np.transpose(y), grid[i].data)
pvgrid[i] = grid[i].data.ravel(order='F')
pyvista2cesium.export(pvgrid, 'my_3D_ model')
```

**Comment by akaszynski** [comment]
@leomiquelutti,

I'm guessing this is due to active normals.  `vtk` will hide the part of mesh that is pointed "inside", but if you want to see the inside of the mesh or the normals are inverted, you can either:
- Disable lighting when plotting with `lighting=False`
- Flip normals with: `flip_normals`

**Comment by leomiquelutti** [comment]
Hi @akaszynski, thanks for the tips! However,

- `lighting=False` didn't work.
- `flip_normals` are only for `PolyData`, right? There are no argument or function `flip_normals` related to `StructuredGrid`, are there?

**Comment by akaszynski** [comment]
Before a mesh is displayed, internally `vtk` extracts the surface mesh and displays that (which is actually a significant part of the plotting time).  What you might try doing is first extracting the surface, flipping the normals, and plotting that.  Try running `extract_surface()` on the structured grid.

**Comment by leomiquelutti** [comment]
Hi @akaszynski.

Although I tried your suggestion, what worked for me was:

1. changing the order of `(x, y, z)` inputs to `(y, z, x)` in the `StructuredGrid` creation. CesiumJS seems to read the exported `glb` file in a different orientation than the one generated with pyvista + [pygltflib](https://pypi.org/project/pygltflib/)
2. `data.ravel('order=C')`
3. `rotate_y(90)`

As in
```
pvgrid = pv.StructuredGrid(y,
    np.transpose(grid[i].data),
    np.transpose(x))
    
pvgrid[i] = grid[i].data.ravel(order='C')
pvgrid.rotate_y(+90)
pvgrid.plot(cmap='jet')
pyvistacesium.export(pvgrid2, 'my_pyvista_model')
```

**Comment by GGDRriedel** [comment]
Can I ask what  "pyvistacesium" is referring to?
Cheers

**Reply by banesullivan** [reply]
Probably the snippet in my original response saved as a separate module: https://github.com/pyvista/pyvista/discussions/5627#discussioncomment-8483135
